### ISSUE TITLE
[Feature] Skip useless computation

### ISSUE THREAD

**Issue by AllentDan** [InternLM/lmdeploy#3663]
### Motivation

I found that the logit in lmdeploy was directly from the last lm_head. Can we skip useless computation from the last attn to lm_head for unwanted logits? llama.cpp applied this for example https://github.com/ggml-org/llama.cpp/blob/5d46babdc2d4675d96ebcf23cac098a02f0d30cc/src/llama-model.cpp#L6888
@lvhan028

**Comment by lvhan028** [comment]
I didn't get it. @grimoire any idea?

**Comment by AllentDan** [comment]
Sorry, the link was outdated. New link is [this](https://github.com/ggml-org/llama.cpp/blob/5d46babdc2d4675d96ebcf23cac098a02f0d30cc/src/llama-model.cpp#L6888).

Prefill for example, we only need the last logit, right? But we still compute all the logits after the last attention. I mean, the last MLP computation for other logits is useless.
